lib/service/connect.go

```diff
@@ -558,19 +556,21 @@ func (process *TeleportProcess) rotate(conn *Connector, localState auth.StateV2,
 	id := conn.ClientIdentity.ID
 	local := localState.Spec.Rotation
 
-	additionalPrincipals, err := process.getAdditionalPrincipals(id.Role)
+	additionalPrincipals, dnsNames, err := process.getAdditionalPrincipals(id.Role)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
+
 	additionalPrincipals = utils.ReplaceInSlice(
 		additionalPrincipals,
 		defaults.AnyAddress,
 		defaults.Localhost,
 	)
 
-	principalsChanged := len(additionalPrincipals) != 0 && !conn.ServerIdentity.HasPrincipals(additionalPrincipals)
+	principalsOrDNSNamesChanged := (len(additionalPrincipals) != 0 && !conn.ServerIdentity.HasPrincipals(additionalPrincipals)) ||
```

Comment:
Btw, I noticed some time ago that if you _remove_ some additional principals (i.e. remove a public addr from config), the change won't be detected and reload will not trigger b/c this checks only for addition. Not sure if that's on purpose or a bug.